src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs

```diff
@@ -1022,10 +1023,21 @@ protected virtual JwtSecurityToken ValidateSignature(string token, TokenValidati
 
             }
 
+            // Get information on where keys used during token validation came from for debugging purposes.
+            // TODO: Once the LKG feature is added to JwtSecurityTokenHandler, this will need to account for the keys
+            // in the Configuration as well.
+            var keysInTokenValidationParameters = TokenUtilities.GetAllSigningKeys(validationParameters);
+            var numKeysInTokenValidationParameters = keysInTokenValidationParameters.Count();
+
             if (kidExists)
             {
                 if (kidMatched)
-                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10511, keysAttempted, jwtToken.Header.Kid, exceptionStrings, jwtToken)));
+                {
+                    var isKidInTVP = keysInTokenValidationParameters.Any(x => x.KeyId.Equals(jwtToken.Header.Kid, StringComparison.Ordinal));
+                    var keyLocation = isKidInTVP ? "TokenValidationParameters" : "Configuration";
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(
+                        LogHelper.FormatInvariant(TokenLogMessages.IDX10511, keysAttempted, LogHelper.MarkAsNonPII(numKeysInTokenValidationParameters), LogHelper.MarkAsNonPII(0), LogHelper.MarkAsNonPII(keyLocation), jwtToken.Header.Kid, exceptionStrings, jwtToken)));
```

Comment:
consider stacking code as above.

Reply:
Fixed! https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1786/commits/27f73791bb1e017be2e7bc1a0b4c25cba31b9a78